Add convenience `get_most_recent` method to fetch newest history metadata with a limit

issammani opened the pull request:
### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->

This patch adds a new convenience method `get_most_recent` to retrieve the newest history metadata entries ordered by updated_at DESC with a configurable limit. Without this method, we have to use `get_since(start: min_i64)` and pop first n elements that we need from the vec which can be wasteful at times. Using `get_most_recent` should give back exactly the number of items we need ordered correctly.

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/releases.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due diligence applied in selecting them.

issammani commented:
cc @mhammond @bendk

mhammond commented on components/places/src/storage/history_metadata.rs:
```diff
@@ -1297,6 +1315,102 @@ mod tests {
+
+        note_observation!(&conn,
+            url "https://example.com/1",
+            view_time Some(10),
```
we should probably have this identical for all visits to make it clear it doesn't matter. Related, this might be racey in that it probably assumes incrementing timestamps between observations, which might not be true, so the order back is different. I'm not immediately sure what to do about that :( Maybe ignore it if other tests rely on it, sqlite doing a full transaction before the timestamp advances seems unlikely to be a problem for a while?

issammani replied:
> Related, this might be racey in that it probably assumes incrementing timestamps between observations, 
Yeah I can see how that can be a problem. I don't see anything in the other tests that attempts to solve it other than maybe some `thread::sleep` that might have been added for this ?

issammani replied:
Sweet ! Done in https://github.com/mozilla/application-services/pull/7002/commits/f8acb29202c6e93b7ca5f15f9cb9ac4ec268e228. I added a tiny helper with a comment. @mhammond if it looks good can you please merge after CI is green. I don't have merge access. Thanks !

mhammond reviewed (commented):
This looks great! I think you will need an addition to the UDL too? And I'm wondering if we should we maybe just replace `get_since()` - it sounds like you are saying it's not useful so it seems likely Android might want the same thing? We wont block on Android changing, but it might be nice to leave a breadcrumb here for future us.

(It's unfortunate this "history-metadata" stuff still exists - it's basically "normal" history with some other "view time" stuff that's not synced - but it is what is it is :)

issammani commented:
> I think you will need an addition to the UDL too?

True. I got so used to the macros now hahah. Added in https://github.com/mozilla/application-services/pull/7002/commits/50eee6bb8cadaf85f4fdd44496410adb35c11e6d

> it sounds like you are saying it's not useful so it seems likely Android might want the same thing? We wont block on Android changing, but it might be nice to leave a breadcrumb here for future us.

Yeah that's the plan. I will also change Android after this lands and we can remove it after. For some reason, I always end up breaking android when I do breaking changes 🙈 So this seems safer. Added ticket and comment on top of method to track this in  https://github.com/mozilla/application-services/pull/7002/commits/d7eaab98d8dbee7ae1b4f28775a41655f1349e06 😄

issammani commented on CHANGELOG.md:
```diff
@@ -3,7 +3,7 @@
 
 ### Swift
-- Added `@unchecked Sendable` to classes that conform to `FeatureManifestInterface`. ([#6963](https://github.com/mozilla/application-services/pull/6963)
+- Added `@unchecked Sendable` to classes that conform to `FeatureManifestInterface`. ([#6963](https://github.com/mozilla/application-services/pull/6963))
```
Was missing a `)` at the end and it bothered me 😂

mhammond reviewed (dismissed):
this is great, thanks!
